### PR TITLE
Generate using TS node types in Typescript Modern

### PR DESCRIPTION
**Pull request by leoasis**
This updates the typescript modern generator to use TS nodes from the `@babel/types` library. This fixes some invalid output in the TS generated files since we were really using flow nodes, which have slight syntax differences.

### CODE REVIEW

File: src/javascript/typescript/codeGeneration.ts [hunk 1]
```diff
@@ -164,19 +163,18 @@ export class TypescriptAPIGenerator extends TypescriptGenerator {
...
-      this.typeAliasObject(operationName, properties)
+      this.interface(operationName, properties)
     );
 
     this.printer.enqueue(exportedTypeAlias);
     this.scopeStackPop();
   }
 
-  public typeAliasesForFragment(fragment: Fragment) {
+  public typesForFragment(fragment: Fragment) {
```
Comment (lewisf) [inline src/javascript/typescript/codeGeneration.ts]: thinking about this now, we might as well call this `interfacesForFragment` since that's actually whats being generated

File: src/javascript/typescript/__tests__/codeGeneration.ts [hunk 2]
```diff
@@ -22,7 +22,7 @@ function compile(
   return compileToIR(schema, document, options);
 }
 
-describe('Typescript codeGeneration', () => {
+describe.only('Typescript codeGeneration', () => {
```
Comment (lewisf) [inline src/javascript/typescript/__tests__/codeGeneration.ts]: i don't think you want this describe.only here

File: src/javascript/typescript/language.ts [hunk 3]
```diff
@@ -69,109 +68,108 @@ export default class TypescriptGenerator {
...
+  public makeNullableType(type: t.TSType) {
+    return t.TSUnionType([
+      type,
+      // t.TSUndefinedKeyword(),
+      t.TSNullKeyword()
     ])
   }
+
+  public isNullableType(type: t.TSType) {
+    return t.isTSUnionType(type) && t.isTSNullKeyword(type.types[1]);
```
Comment (lewisf) [inline src/javascript/typescript/language.ts]: oh cool -- would we want this to check `type.types.some(t.isTSNullKeyword)` just in case?

File: src/javascript/typescript/__tests__/codeGeneration.ts (hunk 2 above)
Comment (leoasis) [reply src/javascript/typescript/__tests__/codeGeneration.ts]: whoops!

File: src/javascript/typescript/language.ts (hunk 3 above)
Comment (leoasis) [reply src/javascript/typescript/language.ts]: Yeah, probably more robust, just didn't do it since we are generating the types ourselves, and the `makeNullableType` method already puts the `null` in the end. But will do it as you suggest

**Comment by leoasis** [comment]
@lewisf regarding the comment about when to use types vs interfaces, the idea I was going with was just using interfaces with "simple" types, which are not part of unions. We could definitely do what you suggest, though. I think it could be useful to have that intermediate type exposed for example to write type guard functions.

File: src/javascript/typescript/codeGeneration.ts (hunk 1 above)
Comment (leoasis) [reply src/javascript/typescript/codeGeneration.ts]: You're right 👍

**Comment by leoasis** [comment]
Alright! I addressed all your comments @lewisf, ready for another look!

**Comment by rosskevin** [comment]
Regarding the unions - here's a real case where I could use a well defined type for each side of the union:

We have a polymorphic type on the backend and on the front where we query (expecting one of the types) it ends up being:

```js
export type LaunchpadListQuery_user_accesses_to =
  | {}
  | {
      id: number
      name: string
      url: string
    }
```
Where one side polymorphic is `{}`.  This means we need to cast any iteration to the type we queried - which in this case I can't reference, only redefine.

TL;DR defining each side of the unions would be a useful enhancement.  Thanks @leoasis

**Comment by leoasis** [comment]
Yeah, already added in the last commit. I agree they are useful to expose, so I added them to the generated output.

**Comment by rosskevin** [comment]
I think I was commenting as you were committing.  You're too quick!

**Comment by lewisf** [comment]
this looks good. really appreciate all your hard work on this @leoasis!
